### ISSUE TITLE
SearchIO blast parser: could not find the query in line

### ISSUE THREAD

**Issue by breviata** [biopython/biopython#554]
I'm running into a problem with the SearchIO blast parser (on either xml or txt outputs).  I'm analyzing thousands of files with 25 blast results per file.  The script works on about 10k sequences and then I get an error on this one file.  If I pass the error, I get about 1000 similar cases of the script failing.  

See below for the error and blast output of the 'problem' sequence along with my code.  Biopython, python and blast are all up-to-date. Be gentle, I'm a biologist turned programmer so most computer speak will be lost on me!!  Thanks for your help!!

**BIOPYTHON ERROR**

```
Traceback (most recent call last):
  File "/home/cstairs/SOFTWARE/MitoPipe_Apr2015_faster2.py", line 463, in <module>
    for blast_record in blast_records:
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/__init__.py", line 316, in parse
    for qresult in generator:
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/BlastIO/blast_text.py", line 38, in __iter__
    for rec in self.blast_iter:
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 1658, in __next__
    return self._parser.parse(StringIO(data))
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 810, in parse
    self._scanner.feed(handle, self._consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 95, in feed
    self._scan_rounds(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 240, in _scan_rounds
    self._scan_alignments(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 392, in _scan_alignments
    self._scan_pairwise_alignments(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 404, in _scan_pairwise_alignments
    self._scan_one_pairwise_alignment(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 421, in _scan_one_pairwise_alignment
    self._scan_hsp(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 456, in _scan_hsp
    self._scan_hsp_alignment(uhandle, consumer)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 487, in _scan_hsp_alignment
    read_and_call(uhandle, consumer.query, start='Query')
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/ParserSupport.py", line 287, in read_and_call
    method(line)
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 1200, in query
    raise ValueError("I could not find the query in line\n%s" % line)
ValueError: I could not find the query in line
Query        ------------------------------------------------------------  
```

*END BIOPYTHON ERROR*

**PROBLEM SEQUENCE OUTPUT**

```
Query= TR11080zzzc0_g2_i2_0

Length=1691
                                                                      Score     E
Sequences producing significant alignments:                          (Bits)  Value

gi|77176709|gb|ABA64469.1|  DNA-dependent RNA polymerase II large...   2156   0.0
gi|702367321|ref|XP_010060951.1|  PREDICTED: DNA-directed RNA pol...   2145   0.0
gi|629102380|gb|KCW67849.1|  hypothetical protein EUGRSUZ_F01570 ...   2137   0.0
gi|475596512|gb|EMT23377.1|  DNA-directed RNA polymerase II subun...   2135   0.0
gi|586648317|ref|XP_006828844.1|  PREDICTED: DNA-directed RNA pol...   2107   0.0
gi|734396220|gb|KHN29539.1|  DNA-directed RNA polymerase II subun...   2104   0.0
gi|685256632|ref|XP_009150077.1|  PREDICTED: DNA-directed RNA pol...   2091   0.0
gi|168066799|ref|XP_001785319.1|  predicted protein [Physcomitrel...   2086   0.0
gi|567216858|ref|XP_006412058.1|  hypothetical protein EUTSA_v100...   2078   0.0
gi|301090036|ref|XP_002895253.1|  DNA-directed RNA polymerase II ...   2078   0.0

*about 1000 sequences....*

gi|465795888|emb|CCV00317.1|  unnamed protein product [Malassezia...   1630   0.0  
gi|237833803|ref|XP_002366199.1|  DNA-directed RNA polymerase II ...   1628   0.0  
gi|50310581|ref|XP_455310.1|  hypothetical protein [Kluyveromyces...   1625   0.0  
gi|590702291|ref|XP_007046588.1|  DNA-directed RNA polymerase II ...   1625   0.0  
gi|590702288|ref|XP_007046587.1|  DNA-directed RNA polymerase II ...   1625   0.0  


>gi|77176709|gb|ABA64469.1| DNA-dependent RNA polymerase II largest subunit [Glaucocystis 
nostochinearum]
Length=1852

 Score =  2156 bits (5587),  Expect = 0.0, Method: Compositional matrix adjust.
 Identities = 1067/1644 (65%), Positives = 1296/1644 (79%), Gaps = 61/1644 (4%)

Query  1     MSAFQFSPIEVKEVSYLQFGVLSPDEMKEMSVCKIEFPQTYENGVPKEGGLSDPRLGTMD  60
             ++ F  S  E+++VS +QFG+L PDE++ MSV KIE+P+ +ENG PK GGL DPR+G++D
Sbjct  4     LTQFPHSSAELRKVSVVQFGILGPDEIRRMSVAKIEYPECFENGRPKLGGLMDPRMGSID  63

Query  61    RTQLCRTCNSDARECPGHFGHIVLAKPMYHVGFLPFVLKILRCVCFFCSKLLVDSNDPRL  120
             R   C+TC     ECPGHFGH+ LAKP+ H+GF+  +LKILRCVCF CSKLL D +D R 
Sbjct  64    RQFKCQTCAGSMSECPGHFGHLELAKPVLHIGFMTTILKILRCVCFHCSKLLCDKSDIRF  123

Query  121   KSILAQNHHRRRIQSMMNLCRTKKICEAGDDADDLAEHETEPEKQRKP-HGGCGNFQPNI  179
             ++ L       R++++M++CR K  CE GD+ D+ AE E  P+ ++K  HGGCGNF P I
Sbjct  124   RNALRLKKPETRLRAVMDICRGKSQCEGGDELDENAEKEIGPDGEKKSRHGGCGNFVPKI  183

Query  180   TKDGLRLLAEFKHVSDESIEKKQVLSAEKVYEILKKITDEDCRMMGLDPKFARPDWMVLT  239
```

*END OUTPUT*

**Excerpt of Script**

```
import glob
import sys
import os
from Bio import SearchIO
from Bio.Seq import Seq
from Bio import SeqIO
import re
import time
import random

for BLAST_file in BLAST_files:
	blast_records = SearchIO.parse(BLAST_file,"blast-text")
	print '$$$$$$ opening file %s $$$$$$$$' % (BLAST_file)
	if Query_num < 25:
		missed = 25 - int(Query_num)
		print 'missed %i sequences in file' % (missed)
	Query_num =0
	print 'trying to parse blast file'
	for blast_record in blast_records:
		Query_num = Query_num + 1
		keep =0
		Query_name = str(blast_record.id)
		print Query_name
		Query_handle = "%s%s" % (FASTA_directory, Query_name)
		Query_file = "%s%s.fasta" % (FASTA_directory, Query_name)
```

**Comment by peterjc** [comment]
The "plain text" BLAST output is notoriously hard to parse - using the latest version of BLAST is often worse because if the NCBI has changed the output slightly Biopython may not cope yet. For this reason we and the NCBI recommend using the XML or tabular output instead of the plain text output.

It looks like you've managed to get an alignment where part of the HSP is just ``------------------------------------------------------------`` for the query, meaning a deletion of at least 60 amino acids relative to the matched protein. Have you already determined this is specific to the query sequence ``TR11080zzzc0_g2_i2_0`` or was that just an example?

Can you post the XML version too? That must give a different result...

P.S.

I edited your formatting using the triple back-tick to make the traceback, sample, and code as literal blocks.

**Comment by breviata** [comment]
Hi Peter, 

Thanks for your help. 

I normally always use xml, but when I got that bug I thought I would try text (as all my Perl friends swear by it). Looking closer I see that I get a different error on a different sequence depending on the xml vs txt output.  On a whim I tried to restrict the blast to only 1 result. No more error.  To me this means that the error comes with one of the subject sequences returned right? That sort of sucks since I want the top 1000 hits!  

The xml error is different (and on a different sequence):

``` File "/home/cstairs/SOFTWARE/MitoPipe_Apr2015_faster2.py", line 463, in <module>
    for blast_record in blast_records:
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/__init__.py", line 316, in parse
    for qresult in generator:
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/BlastIO/blast_xml.py", line 198, in __iter__
    for qresult in self._parse_qresult():
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/BlastIO/blast_xml.py", line 294, in _parse_qresult
    query_id):
  File "/opt/perun/python_packages/lib/python2.7/site-packages/Bio/SearchIO/BlastIO/blast_xml.py", line 395, in _parse_hit
    [a.split(' ', 1) for a in full_id_desc.split(' >')]]
ValueError: need more than 1 value to unpack```

P.S - tried the triple back-tick!! Thanks!

**Comment by breviata** [comment]
Found the error!! BLAST on a gap region - there is no co-ordinate for the query sequence (since there isn't a residue).  I wonder if SearchIO can handle this? Back to XML format I guess - once I deal with that error above... 

Thanks. 

```
Query  1526  SPSSP----FSP--YTPASPGYSPSSPAYSPSSPAYSPTSP-------------------  1560
             SP +P     SP  Y+P SP +S +SP YSP+SP+YSPTSP                   
Sbjct  1566  SPRTPAYGGMSPGAYSPTSPQFSMTSPHYSPTSPSYSPTSPAAGQSPASPSYSPTSPSYS  1625

Query        ------------------------------------------------------------  
                                                                         
Sbjct  1626  PTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPSSPRYSPTSP  1685

Query  1561  AYSPTSPAYSPTSPAYSPTSPAYSPTSPAYS----------PTSPAYSPTSPAYSPTSPA  1610
              YSPTSP YSPTSP YSPTSP YSPTSP+Y           P+SP YSPTSP+YSPTSP 
Sbjct  1686  TYSPTSPTYSPTSPTYSPTSPTYSPTSPSYEGYSPSSPKYSPSSPTYSPTSPSYSPTSPQ  1745
'''

**Comment by peterjc** [comment]
Good work. Could you share the query and match sequence, or just a complete output record from this? I'd like to use it as a test case.

We might want to open a second issue for the XML problem...  CC @bow

**Comment by breviata** [comment]
I'll make another 'issue' for the xml errors tomorrow.  I'm in the process of modifying the source code to figure out what the exact issue is. The blast output is too large... match sequence below.

Here's the query:

```
>TR11080zzzc0_g2_i2_0
MSAFQFSPIEVKEVSYLQFGVLSPDEMKEMSVCKIEFPQTYENGVPKEGGLSDPRLGTMDRTQLCRTCNSDARECPGHFGHIVLAKPMYHVGFLPFVLKILRCVCFFCSKLLVDSNDPRLKSILAQNHHRRRIQSMMNLCRTKKICEAGDDADDLAEHETEPEKQRKPHGGCGNFQPNITKDGLRLLAEFKHVSDESIEKKQVLSAEKVYEILKKITDEDCRMMGLDPKFARPDWMVLTIFPVPPPPVRPSILMDSSSRGEDDLTCKLADIIKSNHALRQQELSGSPAHIITEFTQILQYHVATYLDNELPGLPPAIQRSGRPLKSIRQRLRGKNGRVRGNLMGKRVDFSARTVITPDANIGIDEVGVPRSIALNLTFPDIVTPLNIDRMYEYVRNGPREYPGAKYIVRDDGSRLDLRYIRKPSDLHLDYGYKVERHLRDGDFVLFNRQPSLHKMSIMAHRVKLLPFSTFRLNLSVTSPYNADFDGDEMNLHVPQSLEAIAEAQELMLVPRQIISPQANKPVIGLVQDVLIGARNMTKRDTFIELDTVMNILMCTENFDGRIPMPAILKPKKLWTGKQLFSLILPNVNLIRFTSTHPDGEYTHISPGDTKVLIENGDLISGILCKRTLGTSGGSLIHIICNEHGHDTARLFLNQAQKVVNNWLVNIGFSIGIGDTIADEATMEQINKTIASAKNQVKELVLQAQQNILECQPGRTLHESFENKVNKVLNTARDTAGTSAQNSLKESNNVKSMVTAGSKGSFINISQMIACVGQQNVEGKRIPYGFGHRTLPHFTKDDYGPESRGFVENSYLRGLTPQEFFFHAMGGREGLIDTAVKTSETGYIQRRLVKAMEDLMVRYDGTVRNSLGCIIQFSYGEDGMDGAFVESQKLEILRLGDKAFQTLYFMDPTKANFGEDFLDQEVVEDVRNSSEAYELLSAEYEELKRCRRVLRQEVIPSGDDTWPLPVNLRRLIWNAQVIFRLDTRKSTDLSPVKIIKGVKTLLSRLIVVKGDDEVSLEAQESCTMLFGILLKSTLASKRVLKEFRLNTVAFDWVLGEIESRFMQAIVQPSEAVGAIAAQSIGEPATQMTLNTFHYAGVSSKNVTLGVPRLKELINVAKKVKTPSLTVYLLPHCAKDSERAKSVQCQLEHATLNTVTASTEIFYDPDPTTTVIEEDREFVQAYFEMPDEDISLEKISPWLLRIVLNREMMTDKKLSMADIAEKINIEFGCDTLCIFNDDNAEKLVLHVRIMNDQDPKHEETISEEATTTFLKQLEANMLSEMTLKGIDQIRKVYMREARKTFFDVDGRIATENEWILETDGCNLLQVMSCPEVDFSRTTSNDIVEIIQVLGIEAARAALLKEIRDVISFDGSYVNYRHLAILVDFMTYRGYLMSITRHGINRNVTGPLMRCSFEETVEILMESAAFAEADHLRGVTENIILGQLGKFGTGSFDVFLNEKMLREAVDIPLPDGLEGQELFGDTSPTHQMTPFETMGTPGGAFSPSTPNDGAMFSPFNGYSNEATFSPSGSPSSPFSPYTPASPGYSPSSPAYSPSSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYSPTSPAYTPTSPAYTPTSPAYTPTSPAYTPTSPAYTPTSPAYTPT

```

and subject

```
>gi|308468219|ref|XP_003096353.1| CRE-AMA-1 protein [Caenorhabditis remanei]
MALVGVDFQAPLRTVCRVQFGILGPEEIKRMSVAHVEFPEIYENGKPKLGGLMDPRQGVIDRRGRCMTCA
GNLTDCPGHFAHLELAKPVFHIGFLTKSLKILRCVCFYCGRLLIDKSNPRVIDILKKTSGNAKKRLALIY
DLCKSKSVCEGAAEKEEGLPDDVDDPMSGEKKIPAGCGRYQPSYRRVGIDINAEWKKNVNEDTQERKIML
TAERVLEVFKQITDEDILVIGMDPQFARPEWMICTVLPVPPLAVRPAVVTFGSAKNQDDLTHKLSDIIKT
NQQLQRNEANGAAAHVLTDDVRLLQYHVATLVDNCIPGLPTATQKGGRPLKSIKQRLKGKEGRIRGNLMG
KRVDFSARTVITADPNLPIDTVGVPRTIAQNLTFPEIVTPFNVDKLQELVNRGDTQYPGAKYIIRENGAR
VDLRYHPRAADLHLQPGYRVERHMKDGDIIVFNRQPTLHKMSMMGHRVKILPWSTFRMNLSVTSPYNADF
DGDEMNLHLPQSLETRAEIEEICMVPRQLITPQANKPVMGIVQDTLCAVRMMTKRDVFIDWSFMMDLLMY
LPTWDGKVPQPAILKPKPLWTGKQLFSLIIPGNVNVLRTHSTHPDSEDSGPYKWISPGDTKVLIEHGELL
SGIVCSKTVGKSAGNLLHVVTLELGYEIAANFYSHIQTVINAWLLRVGHTIGIGDTIADHATYLDIQNTI
KKAKQDVVDIIEKAHNDDLEPTPGNTLRQTFENKVNQILNDARDRTGSSAQKSLSEFNNFKSMVVSGSKG
SKINISQVIACVGQQNVEGKRIPFGFRHRTLPHFIKDDYGPESKGFVENSYLAGLTPSEFFFHAMGGREG
LIDTAVKTAETGYIQRRLIKAMESVMVNYDGTVRNSLAQMVQLRYGEDGLDGMWVEDQNMPTMKPNNAVF
ERDFRMDLTDNKFLRKNYSEDVVREIQESEDGISLVESEWSQLEEDRRLLRKIFPRGDAKIVLPCNLQRL
IWNAQKIFKVDLRKPVNLSPLHVINGVRELSKKLIIVSGNDEISKQAQYNATLLMNILLRSTLCTKKMCT
SAKLNTEAFDWLLGEVESRFQQAIAQPGEMVGALAAQSLGEPATQMTLNTFHYAGVSAKNVTLGVPRLKE
IINVSKQLKTPSLTVFLTGAAAKDAEKAKDVLCKLEHTTLKKVTLNTAIYYDPDPKNTVIAEDEEWVSIF
YEMPDHDLSRTSPWLLRIELDRKRMVDKKLTMEMIADRIHGGFGNDVHTIYTDDNAEKLVFRLRIAGEDK
GADTQEEQVDKMEDDVFLRCIEANMLSDLTLQGIPAISKVYMNQPNTDDKKRIIITPEGGFKAVADWILE
TDGTALLRVLAERQIDPVRTTSNDICEIFEVLGIEAVRKAIEREMDNVISFDGSYVNYRHLALLCDVMTA
KGHLMAITRHGINRQEVGALMRCSFEETVDILMEASVHAEVDPVKGVSENIMLGQLARCGTGCFDLVLDV
EKCKYGMEIPQNVVMGAGYYGGFAGSPNAHEFSPSHSPWNSGVTPSYGGASWSPGAGGMSPSAGFSPAGN
MDGGASPFNEGGWSPASPGDPLGALSPRTPAYGGMSPGAYSPTSPQFSMTSPHYSPTSPSYSPTSPAAGQ
SPASPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPTSPSYSPSSPRY
SPTSPTYSPTSPTYSPTSPTYSPTSPTYSPTSPSYEGYSPSSPKYSPSSPTYSPTSPSYSPTSPQYSPTS
PQYSPSSPTYTPSSPTYNPTSPRAFSPQYSPTSPTYSPTSPSYTPSSPQYSPTSPTYTPSPSDQPGTSAQ
YSPSSPTYSPSSPTYSPASPSYSPSSPTYDPNP

```

**Comment by breviata** [comment]
Hi Peter, Please find the xml problem [here](https://github.com/biopython/biopython/issues/556)

Not sure how to proceed on the text problem.... blissful ignorance?
Thanks!

**Comment by peterjc** [comment]
Great - issue #556 for the BLAST XML problem ought to be easier to make headway since you've got a self contained sample output file.

The BLAST text parser has evolved over team with various maintainers tweaking it over the years. We'd really like to deprecate it, but it is still useful and thus far we've been able to keep it working with minor tweaks.

The next step would be to construct a test case giving a problematic plain text output file...

**Comment by breviata** [comment]
Text output is great since its human readable. However I agree that it is incredibly problematic.  I've never used the text parser until this week and was met with nothing but the appearance of more grey hairs.  I'll gladly stick to the xml format (once we sort out the #556 error!!) Thanks again :)

**Comment by peterjc** [comment]
Since you pulled out the sequences, making a test case was actually pretty easy (with the single subject sequence to make a tiny BLAST database):

```
$ blastp -query query.fasta -db subject.fasta -out evil_dash_line.txt
```

```
$ more bug554_searchio.py 
from Bio import SearchIO
for record in SearchIO.parse("evil_dash_line.txt", "blast-text"):
    print record.id
$ python bug554_searchio.py 
/Library/Python/2.7/site-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.
  BiopythonExperimentalWarning)
Traceback (most recent call last):
  File "bug554_searchio.py", line 2, in <module>
    for record in SearchIO.parse("evil_dash_line.txt", "blast-text"):
  File "/Library/Python/2.7/site-packages/Bio/SearchIO/__init__.py", line 316, in parse
    for qresult in generator:
  File "/Library/Python/2.7/site-packages/Bio/SearchIO/BlastIO/blast_text.py", line 38, in __iter__
    for rec in self.blast_iter:
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 1659, in __next__
    return self._parser.parse(StringIO(data))
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 811, in parse
    self._scanner.feed(handle, self._consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 96, in feed
    self._scan_rounds(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 241, in _scan_rounds
    self._scan_alignments(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 393, in _scan_alignments
    self._scan_pairwise_alignments(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 405, in _scan_pairwise_alignments
    self._scan_one_pairwise_alignment(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 422, in _scan_one_pairwise_alignment
    self._scan_hsp(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 457, in _scan_hsp
    self._scan_hsp_alignment(uhandle, consumer)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 488, in _scan_hsp_alignment
    read_and_call(uhandle, consumer.query, start='Query')
  File "/Library/Python/2.7/site-packages/Bio/ParserSupport.py", line 288, in read_and_call
    method(line)
  File "/Library/Python/2.7/site-packages/Bio/Blast/NCBIStandalone.py", line 1201, in query
    raise ValueError("I could not find the query in line\n%s" % line)
ValueError: I could not find the query in line
Query        ------------------------------------------------------------  
```

**Comment by peterjc** [comment]
Work in progress here: https://github.com/peterjc/biopython/tree/evil_dash_line

**Comment by peterjc** [comment]
I think this is fixed now, using your example as a test case - see 23e185375d885a241b0a1d6fbcf615fb4525e8f0 and the preceding commits.

I freely admit this is a hack, but this dash line does seem to be a rare special case. Hopefully this will work for you @breviata - are you able to install Biopython from source and try this?

**Comment by breviata** [comment]
Awesome, thanks everyone!

**Comment by peterjc** [comment]
Assuming that means it works for you @breviata, I'm closing this issue. Please let us know if there are any further problems, and thank you for reporting this and putting together the test case.
